text/2019-02-26-tikv-client-command-line.md

```diff
@@ -0,0 +1,208 @@
+
+JSON was chosen because it is an efficient, pervasive, human-readable, modern
+standard that has many tools (eg `jq`) to use for post processing if needed.
+
+All standard (non-logging) output will be on `stdout`.
+
+It will feature configurable logging levels with `slog` which will output to `stderr`.
+
+The tool will exit zero on successful requests, and non-zero on unsuccessful
+requests. In the case of REPL mode, it shall always exit zero.
+
+### Interacting with it
+
+Users will interact with the tool through the command line. It will support both
+a one-off mode or a REPL (Read-eval-print-loop) style mode (See alternatives).
+
+Example of the one-off raw mode:
+
+```bash
+$ tikv-cli raw set "cncf" "linux foundation"
+Finished in 0.001s.
+$ tikv-cli raw set "tikv" "rust"
+Finished in 0.001s.
+$ tikv-cli raw get "cncf"
+"linux foundation"
+Finished in 0.002s.
+$ tikv-cli raw scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s.
+```
+
+Example of REPL raw mode:
+
+```bash
+$ tikv-cli raw
+> set "cncf" "linux foundation"
+Finished in 0.001s.
+> set "tikv" "rust"
+Finished in 0.001s.
+> get "cncf"
+"linux foundation"
+Finished in 0.001s.
+> scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s.
+```
+
+REPL transactional mode, notice how the `>>` communicates that the user is in a
+transaction. Also notice how the transaction timestamp is outputted in the
+request complete messages.
+
+```bash
+$ tikv-cli txn
+> begin
+Finished in 0.000s. (txn ts: 1551216224838)
+>>set "cncf" "linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> set "tikv" "rust"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> get "cncf"
+"linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+>> scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s. (txn ts: 1551216224838)
+> commit
+Finished in 0.001s. (txn ts: 1551216224838)
+```
+
+One-off transactional mode is a bit different, as we need some way to maintain
+timestamp across commands. The user may choose how to do this, the reccomended
+way is via ENV variables, but they must do this. (See alternatives)
+
+```bash
+$ TS=$(tikv-cli txn begin)
+$ tikv-cli txn ${TS} set "cncf" "linux foundation"
+Finished in 0.001s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} set "tikv" "rust"
+Finished in 0.001s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} get "cncf"
+"linux foundation"
+Finished in 0.002s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s. (txn ts: 1551216224838)
+$ tikv-cli txn ${TS} commit
+Finished in 0.004s. (txn ts: 1551216224838)
+```
+
+## Drawbacks
+
+We may not wish to need to maintain this tool, or we may wish for the community
+to maintain it.

```

Comment:
I think the drawback here is the maintenance burden. We'll need to find some way to handle that. I don't think we need to specify how we handle that here, just be aware that there is this cost.

Reply:
Does the sentence present not communicate this correctly? Because that's exactly what I meant to say. :)